go/vt/vttablet/tabletmanager/rpc_replication.go

```diff
@@ -642,30 +662,81 @@ func (tm *TabletManager) SlaveWasRestarted(ctx context.Context, parent *topodata
 
 // StopReplicationAndGetStatus stops MySQL replication, and returns the
 // current status.
-func (tm *TabletManager) StopReplicationAndGetStatus(ctx context.Context) (*replicationdatapb.Status, error) {
+func (tm *TabletManager) StopReplicationAndGetStatus(ctx context.Context, stopIOThreadOnly bool) (StopReplicationAndGetStatusResponse, error) {
 	if err := tm.lock(ctx); err != nil {
-		return nil, err
+		return StopReplicationAndGetStatusResponse{}, err
 	}
 	defer tm.unlock()
 
-	// get the status before we stop replication
+	// Get the status before we stop replication.
+	// We need to do this first to bail out if the replica has already been stopped.
 	rs, err := tm.MysqlDaemon.SlaveStatus()
 	if err != nil {
-		return nil, vterrors.Wrap(err, "before status failed")
+		return StopReplicationAndGetStatusResponse{}, vterrors.Wrap(err, "before status failed")
 	}
-	if !rs.SlaveIORunning && !rs.SlaveSQLRunning {
-		// no replication is running, just return what we got
-		return mysql.SlaveStatusToProto(rs), nil
-	}
-	if err := tm.stopSlaveLocked(ctx); err != nil {
-		return nil, vterrors.Wrap(err, "stop slave failed")
+	before := mysql.SlaveStatusToProto(rs)
+
+	if stopIOThreadOnly {
+		if !rs.SlaveIORunning {
+			return StopReplicationAndGetStatusResponse{
+				Status: before,
+				Before: before,
+			}, nil
+		}
+		if err := tm.stopSlaveIOThreadLocked(ctx); err != nil {
+			return StopReplicationAndGetStatusResponse{
+				Before: before,
+			}, vterrors.Wrap(err, "stop slave io thread failed")
+		}
+	} else {
+		if !rs.SlaveIORunning && !rs.SlaveSQLRunning {
+			// no replication is running, just return what we got
+			return StopReplicationAndGetStatusResponse{
+				Status: before,
+				Before: before,
```

Comment:
I think we should return `After: before,` as well. If we did nothing, the state after is the same as it was before. Otherwise a caller who wants to know the state after would have to add special logic to look at Before if After is unset.

Reply:
That's a good point. I'll update it.

Reply:
Updated.